### ISSUE TITLE
`clippy::cmp-owned` does not work with nested macros

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#16458]
Using the following flags

`--force-warn clippy::cmp-owned`

this code:

```rust
macro_rules! foo {
    ($($i:ident: $ty:ty, $def:expr);+ $(;)*) => {
        pub fn bar(key: &str) -> bool {
            match key {
                $(
                    stringify!{ first: 1, second: 2 } => {
                        let res = <$ty>::default() == $def;
                        let _i: $ty = $def;
                        return res;
                    }
                )+
                _ => false,
            }
        }
    };
}

foo! {
    max_width: usize, 100;
    required_version: String, env!("HOME").to_string();
}
```

caused the following diagnostics:

```
    Checking _f6f848446eb119812cdc9ef8528fe4617ccc491c v0.1.0 (/tmp/icemaker_global_tempdir.GToAtTsoC3qp/icemaker_clippyfix_tempdir.4a028tID4EfI/_f6f848446eb119812cdc9ef8528fe4617ccc491c)
warning: this creates an owned instance just for comparison
  --> src/lib.rs:20:31
   |
20 |     required_version: String, env!("HOME").to_string();
   |                               ^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `env!("HOME")`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#cmp_owned
   = note: requested on the command line with `--force-warn clippy::cmp-owned`

warning: `_f6f848446eb119812cdc9ef8528fe4617ccc491c` (lib) generated 1 warning (run `cargo clippy --fix --lib -p _f6f848446eb119812cdc9ef8528fe4617ccc491c` to apply 1 suggestion)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.18s

```

However after applying these diagnostics, the resulting code:

```rust
macro_rules! foo {
    ($($i:ident: $ty:ty, $def:expr);+ $(;)*) => {
        pub fn bar(key: &str) -> bool {
            match key {
                $(
                    stringify!{ first: 1, second: 2 } => {
                        let res = <$ty>::default() == $def;
                        let _i: $ty = $def;
                        return res;
                    }
                )+
                _ => false,
            }
        }
    };
}

foo! {
    max_width: usize, 100;
    required_version: String, env!("HOME").to_string();
}
```
no longer compiled:

```
    Checking _f6f848446eb119812cdc9ef8528fe4617ccc491c v0.1.0 (/tmp/icemaker_global_tempdir.GToAtTsoC3qp/icemaker_clippyfix_tempdir.4a028tID4EfI/_f6f848446eb119812cdc9ef8528fe4617ccc491c)
warning: failed to automatically apply fixes suggested by rustc to crate `_f6f848446eb119812cdc9ef8528fe4617ccc491c`

after fixes were automatically applied the compiler reported errors within these files:

  * src/lib.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust-clippy/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0308]: mismatched types
  --> src/lib.rs:20:31
   |
20 |     required_version: String, env!("HOME");
   |                       ------  ^^^^^^^^^^^^ expected `String`, found `&str`
   |                       |
   |                       expected due to this
   |
help: try using a conversion method
   |
20 |     required_version: String, env!("HOME").to_string();
   |                                           ++++++++++++

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0308`.
Original diagnostics will follow.

warning: this creates an owned instance just for comparison
  --> src/lib.rs:20:31
   |
20 |     required_version: String, env!("HOME").to_string();
   |                               ^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `env!("HOME")`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#cmp_owned
   = note: requested on the command line with `--force-warn clippy::cmp-owned`

warning: `_f6f848446eb119812cdc9ef8528fe4617ccc491c` (lib) generated 1 warning (run `cargo clippy --fix --lib -p _f6f848446eb119812cdc9ef8528fe4617ccc491c` to apply 1 suggestion)
       Fixed src/lib.rs (1 fix)
warning: `_f6f848446eb119812cdc9ef8528fe4617ccc491c` (lib test) generated 1 warning (1 duplicate)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.37s

```

Version:
```
rustc 1.95.0-nightly (75963ce79 2026-01-25)
binary: rustc
commit-hash: 75963ce795666bc1f961e5d60060809809f6bc68
commit-date: 2026-01-25
host: x86_64-unknown-linux-gnu
release: 1.95.0-nightly
LLVM version: 21.1.8

```
